apps/arweave/src/ar_node_worker.erl

```diff
@@ -1926,3 +1710,317 @@ dump_mempool(TXs, MempoolSize) ->
+	#mining_solution{
+		last_step_checkpoints = LastStepCheckpoints,
+		mining_address = MiningAddress,
+		next_seed = NonceLimiterNextSeed,
+		next_vdf_difficulty = NonceLimiterNextVDFDifficulty,
+		nonce = Nonce,
+		nonce_limiter_output = NonceLimiterOutput,
+		partition_number = PartitionNumber,
+		poa1 = PoA1,
+		poa2 = PoA2,
+		preimage = SolutionPreimage,
+		recall_byte1 = RecallByte1,
+		recall_byte2 = RecallByte2,
+		solution_hash = SolutionH,
+		start_interval_number = IntervalNumber,
+		step_number = StepNumber,
+		steps = SuppliedSteps
+	} = Solution,
+	MerkleRebaseThreshold = ?MERKLE_REBASE_SUPPORT_THRESHOLD,
+
+	#block{ indep_hash = PrevH, timestamp = PrevTimestamp,
+			wallet_list = WalletList,
+			nonce_limiter_info = PrevNonceLimiterInfo } = PrevB,
+	Now = os:system_time(second),
+	MaxDeviation = ar_block:get_max_timestamp_deviation(),
+	Timestamp =
+		case Now < PrevTimestamp - MaxDeviation of
+			true ->
+				?LOG_WARNING([{event, clock_out_of_sync},
+						{previous_block, ar_util:encode(PrevH)},
+						{previous_block_timestamp, PrevTimestamp},
+						{our_time, Now},
+						{max_allowed_deviation, MaxDeviation}]),
+				PrevTimestamp - MaxDeviation;
+			false ->
+				Now
+		end,
+
+	IsBanned = ar_node_utils:is_account_banned(MiningAddress,
+			ar_wallets:get(WalletList, MiningAddress)),
+
+	%% Check the solution is ahead of the previous solution on the timeline.
+	NonceLimiterInfo = #nonce_limiter_info{ global_step_number = StepNumber,
+			output = NonceLimiterOutput,
+			prev_output = PrevNonceLimiterInfo#nonce_limiter_info.output },
+	PassesTimelineCheck =
+		case IsBanned of
+			true ->
+				ar_events:send(solution, {rejected, #{ reason => mining_address_banned,
+						source => Source }}),
+				{false, address_banned};
+			false ->
+				case ar_nonce_limiter:is_ahead_on_the_timeline(NonceLimiterInfo,
+						PrevNonceLimiterInfo) of
+					false ->
+						ar_events:send(solution, {stale, #{ source => Source }}),
+						{false, timeline};
+					true ->
+						true
+				end
+		end,
+
+	%% Check solution seed.
+	#nonce_limiter_info{ next_seed = PrevNextSeed,
+			next_vdf_difficulty = PrevNextVDFDifficulty,
+			global_step_number = PrevStepNumber } = PrevNonceLimiterInfo,
+	PrevIntervalNumber = PrevStepNumber div ?NONCE_LIMITER_RESET_FREQUENCY,
+	PassesSeedCheck =
+		case PassesTimelineCheck of
+			{false, Reason} ->
+				{false, Reason};
+			true ->
+				case {IntervalNumber, NonceLimiterNextSeed, NonceLimiterNextVDFDifficulty}
+						== {PrevIntervalNumber, PrevNextSeed, PrevNextVDFDifficulty} of
+					false ->
+						ar_events:send(solution, {stale, #{ source => Source }}),
+						{false, seed_data};
+					true ->
+						true
+				end
+		end,
+
+	%% Check solution difficulty
+	PrevDiffPair = ar_difficulty:diff_pair(PrevB),
+	LastRetarget = PrevB#block.last_retarget,
+	PrevTS = PrevB#block.timestamp,
+	DiffPair = {_PoA1Diff, Diff} = ar_retarget:maybe_retarget(PrevB#block.height + 1,
```

Comment:
we can't use `get_current_diff` any more because if we're switching to the sibling tip the difficulty cached in `node_state` may not be correct?

Reply:
yep